### PR TITLE
[flutter_plugin_tools] Migrate 'test' to new base command

### PR DESCRIPTION
**Pull request by stuartmorgan**
Migrates `test` to the new package looping base command.

Refactors the bulk of the logic to helpers for easier understanding of the flow.

Part of flutter/flutter#83413

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran [the auto-formatter]. (Note that unlike the flutter/flutter repo, the flutter/plugins repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I [updated pubspec.yaml](https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates) with an appropriate new version according to the [pub versioning philosophy].
- [ ] I updated CHANGELOG.md to add a description of the change.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test exempt.
- [x] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[the auto-formatter]: https://github.com/flutter/plugins/blob/master/script/tool/README.md#format-code

### CODE REVIEW

**Comment by stuartmorgan** [comment]
This is probably the simplest command we have, and it hadn't been migrated, so that works out well as an intro to these refactors :) I think the only context you'll need here is to look at https://github.com/flutter/plugins/blob/master/script/tool/lib/src/common/package_looping_command.dart

Note that this is divided into two commits: the first is just a simple local refactor to pull the bulk of the logic into helpers and do some minor cleanup on them (e.g., using `if` in lists instead of the older method of conditional `add` calls in more places). The second is the migration itself, and should be very simple to follow given the first commit (I originally did them the other way, but the diffs weren't very good due to the whitespace changes + cleanup).

File: script/tool/lib/src/test_command.dart [hunk 1]
```diff
@@ -32,72 +32,64 @@ class TestCommand extends PluginCommand {
...
-          if (enableExperiment.isNotEmpty)
-            '--enable-experiment=$enableExperiment',
-        ];
+    final int exitCode = await processRunner.runAndStream(
+      flutterCommand,
+      <String>[
+        'test',
+        '--color',
+        if (experiment.isNotEmpty) '--enable-experiment=$experiment',
+        if (isWebPlugin(package)) '--platform=chrome',
```
Comment (ditman) [inline script/tool/lib/src/test_command.dart:59]: Can we add a deprecation warning to this, so people stop doing flutter test --platform=chrome? Ideally, this should be deleted once the last couple of plugins are migrated to flutter drive (soon)

File: script/tool/lib/src/test_command.dart [hunk 2]
```diff
@@ -32,72 +32,64 @@ class TestCommand extends PluginCommand {
...
-      if (exitCode != 0) {
-        failingPackages.add(packageName);
-      }
+  /// Runs the Dart tests for a non-Flutter package, returning true on success.
+  Future<bool> _runDartTests(Directory package) async {
+    // Unlike `flutter test`, `pub run test` does not automatically get
+    // packages
+    int exitCode = await processRunner.runAndStream(
+      'dart',
+      <String>['pub', 'get'],
```
Comment (ditman) [inline script/tool/lib/src/test_command.dart:72]: (Isn't it slightly odd that flutter tests don't attempt to `flutter pub get` too?)

File: script/tool/lib/src/test_command.dart [hunk 3]
```diff
@@ -20,7 +19,8 @@ class TestCommand extends PluginCommand {
     argParser.addOption(
       kEnableExperiment,
```
Comment (ditman) [inline script/tool/lib/src/test_command.dart:20]: What is `kEnableExperiment`? Is it a comma-separated value? or a single experiment? `valueHelp` might be useful here too!

File: script/tool/lib/src/test_command.dart [hunk 4]
```diff
@@ -32,72 +32,64 @@ class TestCommand extends PluginCommand {
...
-      if (!packageDir.childDirectory('test').existsSync()) {
-        print('SKIPPING $packageName - no test subdirectory');
-        continue;
-      }
+  Future<PackageResult> runForPackage(Directory package) async {
+    if (!package.childDirectory('test').existsSync()) {
+      return PackageResult.skip('No test/ directory.');
+    }
 
-      print('RUNNING $packageName tests...');
```
Comment (ditman) [inline script/tool/lib/src/test_command.dart:45]: There's a lot of missing "reporting". Is that done by the PackageLoopingCommand now? (not that I'm going to miss it, just making sure that the output log stays useful)

**Review by ditman — APPROVED** [review]
This is a great refactor. The code is wwwway simpler to follow now!

File: script/tool/lib/src/test_command.dart (hunk 2 above)
Comment (stuartmorgan) [reply script/tool/lib/src/test_command.dart:75]: See the comment just above this line.

File: script/tool/lib/src/test_command.dart (hunk 3 above)
Comment (stuartmorgan) [reply script/tool/lib/src/test_command.dart:20]: Added a link like the one in `flutter`'s help.

File: script/tool/lib/src/test_command.dart (hunk 4 above)
Comment (stuartmorgan) [reply script/tool/lib/src/test_command.dart:45]: Standardized output for starting each package's run is part of the base command, yes.

File: script/tool/lib/src/test_command.dart (hunk 1 above)
Comment (stuartmorgan) [reply script/tool/lib/src/test_command.dart:62]: Added a comment.

**Pull request merged**